csrc/ir_utils.cpp

```diff
@@ -572,6 +574,83 @@ bool isReductionTvOp(const Expr* expr) {
   return ir_utils::isTvOp(expr) && isReductionOp(expr);
 }
 
+bool isReductionOverSizeZero(
+    const Expr* expr,
+    SchedulerRuntimeInfo& runtime_info) {
+  if (!isReductionOp(expr)) {
+    return false;
+  }
+  auto expr_eval = runtime_info.expressionEvaluator();
+  for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
+    for (auto id : output->getRootDomain()) {
+      if (id->isReduction() &&
+          expr_eval.evaluate(id->extent())->as<int64_t>() == 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool reductionHasSizeZeroConcreteDimension(
+    const Expr* expr,
+    SchedulerRuntimeInfo& runtime_info) {
+  if (!isReductionOp(expr)) {
+    return false;
+  }
+  auto expr_eval = runtime_info.expressionEvaluator();
+  for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
+    for (auto id : output->getRootDomain()) {
+      if (!id->isReduction() &&
+          expr_eval.evaluate(id->extent())->as<int64_t>() == 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void replaceReductionWithFull(Expr* expr) {
+  auto fusion = expr->container()->as<Fusion>();
+  if (expr->isA<ReductionOp>()) {
+    // Fill value is reduction starting value
+    auto rop = expr->as<ReductionOp>();
+    auto old_out = rop->out()->as<TensorView>();
+    auto new_out = full_like(old_out, rop->init());
+    ir_utils::replaceValue(fusion, {{old_out, new_out}});
```

Comment:
This may be fine, but would like to confirm if my understanding is correct.

`replaceValue` just does replacements based on the given mappings. I'm not 100% sure, but this seems like it'd apply a replacement to the reduction op, resulting in a new reduction op from the same input to `new_out`. The full op would still be there. All the uses of `old_out` would be replaced with `new_out`, so in the end I think this would result in what we would like, but the replacement of the reduction op should be avoided (if it actually happened) since the new output would not be a valid reduction output tensor.

Reply:
I'd think this would be better:

```
for (auto use: old_out->uses()) {
  ir_utils::replaceValInExpr(use, old_out, new_out);
}
```

Reply:
I think you're saying the only difference between using `ir_utils::replaceValue` and your proposed loop is that the former would change `rop` to have an incorrect output, and the latter wouldn't touch `rop`.

This is because even though `rop` shouldn't be used since it's disconnected from the outputs, it's still not great to have an illegally defined reduction operation hanging out.